### PR TITLE
dynamodocstore: preliminary implementation for Amazon DynamoDB

### PR DESCRIPTION
**Pull request by jba**

### CODE REVIEW

**Comment by vangent** [comment]
I'm happy to let Tuo review this ;-).

File: internal/docstore/dynamodocstore/example_test.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"gocloud.dev/internal/docstore/dynamodocstore"
+)
+
+func Example() {
+	ctx := context.Background()
+	sess, err := session.NewSession()
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/example_test.go]: My shell has
```
export AWS_REGION=us-east-2
```

[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/example_test.go]: Do all our examples document all the needed config? That would be a lot of duplication.

File: internal/docstore/dynamodocstore/codec.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+func encodeDoc(doc driver.Document) (map[string]*dyn.AttributeValue, error) {
+	var e encoder
+	if err := doc.Encode(&e); err != nil {
+		return nil, err
+	}
+	return e.av.M, nil
+}
+
+// TODO: improve driver.Encoder to make this efficient.
+func encodeDocKeyFields(doc driver.Document, pkey, skey string) (map[string]*dyn.AttributeValue, error) {
```
Comment (shantuo) [inline internal/docstore/dynamodocstore/codec.go]: Could you add some documentation to this function? It was not super obvious what's going on here, expecially what the arguments are.

File: internal/docstore/dynamodocstore/example_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/example_test.go]: Oh, I see, I had an Output comment. Removed it.

File: internal/docstore/dynamodocstore/codec.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+var nullValue = new(dyn.AttributeValue).SetNULL(true)
+
+type encoder struct {
+	av *dyn.AttributeValue
+}
+
+func (e *encoder) EncodeNil()            { e.av = nullValue }
+func (e *encoder) EncodeBool(x bool)     { e.av = new(dyn.AttributeValue).SetBOOL(x) }
+func (e *encoder) EncodeInt(x int64)     { e.av = new(dyn.AttributeValue).SetN(strconv.FormatInt(x, 10)) }
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/codec.go]: It's basically the same thing, and I prefer the symmetry with strconv.FormatUint.

File: internal/docstore/dynamodocstore/codec.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+// TODO: improve driver.Encoder to make this efficient.
+func encodeDocKeyFields(doc driver.Document, pkey, skey string) (map[string]*dyn.AttributeValue, error) {
+	m, err := encodeDoc(doc)
+	if err != nil {
+		return nil, err
+	}
+	hasP := false
+	hasS := false
+	if skey == "" {
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/codec.go]: Weird function. It took me a while to remember what it does. I hope the comment explains it clearly now.

File: internal/docstore/dynamodocstore/dynamo_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+func (h *harness) Close() {
+	h.closer()
+}
+
+func (h *harness) MakeCollection(context.Context) (driver.Collection, error) {
+	return newCollection(dyn.New(h.sess), collectionName, keyName, ""), nil
+}
+
+func TestConformance(t *testing.T) {
+	clearTable(t)
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/dynamo_test.go]: Done.

File: internal/docstore/dynamodocstore/dynamo.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+	dyn "github.com/aws/aws-sdk-go/service/dynamodb"
+	"gocloud.dev/internal/docstore"
+	"gocloud.dev/internal/docstore/driver"
+	"gocloud.dev/internal/gcerr"
+)
+
+type collection struct {
+	db           *dyn.DynamoDB
+	name         string // DynamoDB table name
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/dynamo.go]: Done.

File: internal/docstore/dynamodocstore/dynamo.go [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+		},
+	}
+	return c
+
+}
+
+func (c *collection) KeyFields() []string {
+	if c.sortKey == "" {
+		return []string{c.partitionKey}
+	} else {
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/dynamo.go]: Done.

File: internal/docstore/dynamodocstore/dynamo.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+		if err != nil {
+			return i, err
+		}
+	}
+	return len(actions), nil
+}
+
+func (c *collection) missingKeyField(m map[string]*dyn.AttributeValue) string {
+	if _, ok := m[c.partitionKey]; !ok {
+		return c.partitionKey
```
[the remark this answers is not in the document]
Comment (jba) [reply internal/docstore/dynamodocstore/dynamo.go]: Both can be missing, but if the partition key is missing it's not even worth mentioning the sort key.

File: internal/docstore/dynamodocstore/codec.go (hunk 2 above)
Comment (jba) [reply internal/docstore/dynamodocstore/codec.go]: Done in response to earlier comment.

**Comment by jba** [comment]
PTAL.

**Pull request merged**